### ISSUE TITLE
Result Mismatch with the original results in paper in human faces

### ISSUE THREAD

**Issue by mu-cai** [rosinality/swapping-autoencoder-pytorch#6]
Hi Rosinality,

Thanks for your excellent code! It is quite excellent!
However, I found that the result of human images is not good. For example, your results are:
![image](https://user-images.githubusercontent.com/65588556/89730668-78bd7d80-da73-11ea-97a8-2f19aaa25217.png)

(It is obvious that the eyeglasses for the generated images are different from the paper's claim.)
![image](https://user-images.githubusercontent.com/65588556/89730674-87a43000-da73-11ea-84c8-1782b2f18a61.png)

Maybe the problem is the training scheme? Or the cropping method?

**Comment by rosinality** [comment]
Yes, in mumy cases the effect of the texture code in quite restrictive. I suspect that maybe the differences of training scheme was crucial, but I don't know what could be. (Maybe it is related to image/patch resolutions.)

**Comment by mu-cai** [comment]
@rosinality 
Thanks for your information!
During the past few days, I have trained your model under CeleAHQ dataset, and the result is quite bad.
![image](https://user-images.githubusercontent.com/65588556/90418758-24a44000-e0e8-11ea-96f9-c72e90f31028.png)
each row means original A, original B, reconstructed A and structure A + structure B. You can see that the resulting image even can't keep the pose of A.

Also, I trained it on the LSUN church dataset, the result is also not good.

![image](https://user-images.githubusercontent.com/65588556/90419149-ba3fcf80-e0e8-11ea-9c23-6b9a30a594b5.png)

I think there may be several possible issues:

(1) padding, someone has already pointed that out.
(2) The crop. As for the church dataset, they don't do the resize first, instead, they do the crop.

![image](https://user-images.githubusercontent.com/65588556/90419335-03901f00-e0e9-11ea-8174-c3937e519517.png)

I also wonder how can they keep the original image size when keeping the ratio and the short side(256) unchanged.

(3) The co-occur unit. You can see that in paper they stated that 
![image](https://user-images.githubusercontent.com/65588556/90419501-3d612580-e0e9-11ea-9207-d34e494bd6cb.png)

And for each prediction, they did the following operation:

![image](https://user-images.githubusercontent.com/65588556/90419564-57026d00-e0e9-11ea-841d-e68778d717be.png)
 
So this operation should be down for 8 times.

Thanks again for your nice work!

Best,
Mu

**Comment by rosinality** [comment]
1. Could you let me know which part of the padding is incorrect?
2. LSUN church dataset is already resized to be shorter side size is 256, so resizing will not affect the result.
3. Do you mean that cooccur discriminator should be done on 8 patches? Hmm, it seems like that this is different from the paper. I will try to fix this.

Thank you for your testing!

**Comment by mu-cai** [comment]
@rosinality 

Thanks for your reply!
1. You have already fixed this problem yesterday(already committed).
2. Yes, the shorter side is 256, however, the size of the longer side is not fixed. However, during training, you resized the image into a square in your code, making the ratio of two sides changed, which mismatched with the paper.
![image](https://user-images.githubusercontent.com/65588556/90456207-eed47b00-e12a-11ea-8029-c85778711bf1.png)

3. Yes! Your single operation should be done for 8 times. Because when you sample a patch from one real image and 8 patches from the fake image, you will get just one prediction. You need 8$N$  predictions, no $N$.

Thanks again for your answer!

Best,
Mu

**Comment by rosinality** [comment]
1. Actually padding will not affect results, as that bug will only affect 1x1 convs in current implementation
2. As prepare_data.py will do resizing with torchvision, it will respect aspect ratios by default.
3. Seems like that it is important issue. Fixed it at 38cb3aeecbc232a1c1a405c45f1026bd5db36061.

**Comment by mu-cai** [comment]
@rosinality 

Thanks for your quick programming!
I have run your code just now, and one more question:
In your code, for each structure/texture pair, you have 8 crops for the real/fake images, but only 4 crops for the ref image. However, I think that for each crop of the real/fake image, we need 4 patches. That is to say, we need 4*8=32 patches in total.

This is my understanding, however, the author didn't state this in his paper... what is your opinion?

Mu

**Comment by rosinality** [comment]
Hmm maybe you are right. But as model uses mean of ref image vectors, maybe it is not very different from using distinct ref patches for each samples. (Hopefully.)

**Comment by rosinality** [comment]
Anyway, I have changed to use distinct reference samples for each samples. It is less resource consuming that I thought.

**Comment by mu-cai** [comment]
@rosinality 

Thanks for your working! Yes, in my opinion, if the training iterations are large enough, then the fixed reference samples would produce the same result as the distinct reference samples. Yes, I also think that the model would be more robust if adopting the distinct reference patches. The GPU memory won't increase too much if doing so... also superised.

Mu

**Comment by zhangqianhui** [comment]
My tf implementation: https://github.com/zhangqianhui/Swapping-Autoencoder-tf. Hope to help you.

**Comment by gilevir** [comment]
Hi @mu-cai,

Did the above corrections led to better structure/style swapping results on your side ?
